src/infra/core/src/utils/simple_transfer_protocol.rs

```diff
@@ -214,6 +240,143 @@ impl SimpleTransferProtocol {
+            .as_data_repo()
+            .get_stream(&data_slice.physical_hash)
+            .await
+        {
+            Ok(s) => Ok(s),
+            Err(GetError::NotFound(e)) => Err(CorruptedSourceError {
+                message: "Source data file is missing".to_owned(),
+                source: Some(e.into()),
+            }
+            .into()),
+            Err(GetError::Access(e)) => Err(SyncError::Access(e)),
+            Err(GetError::Internal(e)) => Err(SyncError::Internal(e)),
+        }?;
+
+        match dst
+            .as_data_repo()
+            .insert_stream(
+                stream,
+                InsertOpts {
+                    precomputed_hash: if !trust_source_hashes {
+                        None
+                    } else {
+                        Some(&data_slice.physical_hash)
+                    },
+                    expected_hash: Some(&data_slice.physical_hash),
+                    size_hint: Some(data_slice.size as usize),
+                    ..Default::default()
+                },
+            )
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(InsertError::HashMismatch(e)) => Err(CorruptedSourceError {
+                message: concat!(
+                    "Data file hash declared by the source didn't match ",
+                    "the computed - this may be an indication of hashing ",
+                    "algorithm mismatch or an attempted tampering",
+                )
+                .to_owned(),
+                source: Some(e.into()),
+            }
+            .into()),
+            Err(InsertError::Access(e)) => Err(SyncError::Access(e)),
+            Err(InsertError::Internal(e)) => Err(SyncError::Internal(e)),
+        }?;
+
+        stats.src.data_slices_read += 1;
+        stats.dst.data_slices_written += 1;
+        stats.src.bytes_read += data_slice.size as usize;
+        stats.dst.bytes_written += data_slice.size as usize;
+        listener.on_status(SyncStage::TransferData, &stats);
+
+        Ok(())
+    }
+
+    async fn downlaod_block_checkpoints<'a>(
+        &'a self,
+        src: &'a dyn Dataset,
+        dst: &'a dyn Dataset,
+        checkpoint: &Checkpoint,
+        trust_source_hashes: bool,
+        listener: Arc<dyn SyncListener>,
+        mut stats: SyncStats,
+    ) -> Result<(), SyncError> {
+        tracing::info!(hash = ?checkpoint.physical_hash, "Transfering checkpoint file");
+
+        let stream = match src
+            .as_checkpoint_repo()
+            .get_stream(&checkpoint.physical_hash)
+            .await
+        {
+            Ok(s) => Ok(s),
+            Err(GetError::NotFound(e)) => Err(CorruptedSourceError {
+                message: "Source checkpoint file is missing".to_owned(),
+                source: Some(e.into()),
+            }
+            .into()),
+            Err(GetError::Access(e)) => Err(SyncError::Access(e)),
+            Err(GetError::Internal(e)) => Err(SyncError::Internal(e)),
```

Comment:
Proposition: extract error mapping logic into `impl From<GetError> for SyncError`

For example, like `impl From<ContainsError> for GetError`